atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java

```diff
@@ -1677,43 +1679,89 @@ public void addGarbageCollectionSentinelValues(TableReference tableRef, Set<Cell
-                                rowName,
-                                tableName,
-                                ImmutableList.of(),
-                                ImmutableList.of(col),
-                                ConsistencyLevel.SERIAL,
-                                writeConsistency));
-                        if (!casResult.isSuccess()) {
-                            throw new KeyAlreadyExistsException("This transaction row already exists.",
-                                    ImmutableList.of(e.getKey()));
-                        }
+            clientPool.runWithRetry(client -> {
+                for (Entry<Cell, byte[]> e : values.entrySet()) {
+                    CheckAndSetRequest request = CheckAndSetRequest.newCell(tableRef, e.getKey(), e.getValue());
+                    CASResult casResult = executeCheckAndSet(client, request);
+                    if (!casResult.isSuccess()) {
+                        throw new KeyAlreadyExistsException("This transaction row already exists.",
+                                ImmutableList.of(e.getKey()));
                     }
-                    return null;
                 }
+                return null;
             });
         } catch (Exception e) {
             throw Throwables.throwUncheckedException(e);
         }
     }
 
     @Override
+    public void checkAndSet(final CheckAndSetRequest request) throws CheckAndSetException {
+        try {
+            clientPool.runWithRetry(client -> {
+                CASResult casResult = executeCheckAndSet(client, request);
+
+                if (!casResult.isSuccess()) {
+                    List<byte[]> currentValues = casResult.current_values.stream()
+                            .map(Column::getValue)
+                            .collect(Collectors.toList());
+
+                    String msg = String.format("The cell %s in table %s has an unexpected value. "
+                            + "If this is happening repeatedly, your program may be out of sync with the database.",
+                            request.cell(),
+                            request.table().getQualifiedName());
+
+                    throw new CheckAndSetException(msg,
+                            request.cell(),
+                            request.oldValue().orElse(null),
+                            currentValues);
+                }
+                return null;
+            });
+        } catch (TException e) {
+            throw Throwables.throwUncheckedException(e);
+        }
+    }
+
+    private CASResult executeCheckAndSet(Client client, CheckAndSetRequest request)
+            throws TException {
+        TableReference table = request.table();
+        Cell cell = request.cell();
+        long timestamp = AtlasDbConstants.TRANSACTION_TS;
+
+        ByteBuffer rowName = ByteBuffer.wrap(cell.getRowName());
+        byte[] colName = CassandraKeyValueServices
```

Comment:
So `makeCompositeBuffer` returns a buffer, but we then need to pass the underlying array into `makeColumn` in two places below, so I think it makes sense to keep it as a `byte[]` here.